velocitas_lib/velocitas_lib.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) 2023 Robert Bosch GmbH
+#
+# This program and the accompanying materials are made available under the
+# terms of the Apache License, Version 2.0 which is available at
+# https://www.apache.org/licenses/LICENSE-2.0.
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import os
+import re
+import sys
+from pathlib import Path
+from typing import Any
+
+
+def require_env(name: str) -> str:
+    """Require and return an environment variable.
+
+    Args:
+        name (str): The name of the variable.
+
+    Raises:
+        ValueError: In case the environment variable is not set.
+
+    Returns:
+        str: The value of the variable.
+    """
+    var = os.getenv(name)
+    if not var:
+        raise ValueError(f"Environment variable {name!r} not set!")
+    return var
+
+
+def get_workspace_dir() -> str:
+    """Return the workspace directory."""
+    return require_env("VELOCITAS_WORKSPACE_DIR")
+
+
+def get_app_manifest() -> dict:
+    return json.loads(require_env("VELOCITAS_APP_MANIFEST"))
+
+
+def get_script_path() -> str:
+    """Return the absolute path to the directory the invoked Python script
+    is located in."""
+    return os.path.dirname(os.path.realpath(sys.argv[0]))
+
+
+def get_package_path() -> Path:
+    """Return the absolute path to the package directory the invoked Python script
+    is located in."""
+    return Path(__file__).resolve().parents[1]
+
+
+def get_cache_data() -> dict[str, Any]:
+    """Return the data of the cache as Python object."""
+    return json.loads(require_env("VELOCITAS_CACHE_DATA"))
+
+
+def get_services() -> dict[str, Any]:
+    """Return all specified services as Python object."""
+    return json.load(
+        open(
+            f"{Path(get_script_path()).resolve().parents[1]}/runtime.json",
+            encoding="utf-8",
+        )
+    )
+
+
+def replace_variables(input_str: str, variables: dict[str, str]) -> str:
+    """Replace all occurrences of the defined variables in the input string"""
+    if "${{" not in input_str:
+        return input_str
+    input_str_match = re.search(r"(?<=\${{)(.*?)(?=}})", input_str)
```

Comment:
No heavy regex search needed. Simply iterate over all key, value pairs in the `variables` dict and do a replace all occurrences which match the pattern:

```python
for key, value in variables.items():
    input_str = input_str.replace('${{ ' + key + ' }}', value)
```

Reply:
when and how to raise errors then? It was requested to have error handling on specific conditions

Reply:
Looks like you had a discussion about it already. Let's keep it for now.